parTraverseN and parSequenceN

TimWSpence opened the pull request:
Stop-gap solution while we figure out how we want to do this for any `Concurrent[F]`. This requires us to decide whether we want to move `Queue` or `Semaphore` into `kernel` - see https://github.com/typelevel/cats-effect/issues/1291 for more details

TimWSpence committed:
Port parTraverseN and parSequenceN

TimWSpence committed:
Ambiguous implicit

djspiewak commented on core/shared/src/main/scala/cats/effect/IO.scala:
```diff
@@ -280,6 +281,24 @@ object IO extends IOCompanionPlatform with IOLowPriorityImplicits {
+  /**
+   * Like `Parallel.parTraverse`, but limits the degree of parallelism.
+   */
+  def parTraverseN[T[_]: Traverse, A, B](n: Long)(ta: T[A])(f: A => IO[B]): IO[T[B]] =
+    for {
+      semaphore <- Semaphore[IO](n)
+      tb <- ta.parTraverse { a => semaphore.permit.use(_ => f(a)) }
+    } yield tb
```
I would prefer we use the `Queue` implementation (as @Avasil suggested) rather than the `Semaphore` variant.

TimWSpence committed:
Move Semaphore to kernel

So we can have a default implementation of parTraverseN and parSequenceN
for Concurrent

TimWSpence committed:
Provide a default implementation of par{Sequence, Traverse}N in terms of Semaphore

TimWSpence committed:
Default implementation of par{Sequence, Traverse}N for Concurrent

TimWSpence committed:
Typeclass to allow par{Sequence, Traverse}N to be over-ridden

TimWSpence committed:
Remove ParTraverseN typeclass

TimWSpence committed:
Override parSequenceN for IO, based on Monix queue implementation

TimWSpence committed:
Always cancel fibers in parSequenceN

TimWSpence committed:
Special case parTraverseN for IO

RaasAhsan commented on core/shared/src/main/scala/cats/effect/IO.scala:
```diff
@@ -277,6 +278,53 @@ object IO extends IOCompanionPlatform with IOLowPriorityImplicits {
+              .take
+              .flatMap {
+                case (p, task) =>
+                  task.redeemWith(
```
interesting question: should we spawn this task in another fiber so that self-cancellation doesn't interrupt a worker? then we can respond properly by terminating the whole operation

RaasAhsan commented on core/shared/src/main/scala/cats/effect/IO.scala:
```diff
@@ -277,6 +278,53 @@ object IO extends IOCompanionPlatform with IOLowPriorityImplicits {
+          case Right(values) =>
+            IO.pure(values)
+        }
+      } { fibers => fibers.traverse(_.cancel).void }
```
We can cancel these fibers in parallel

RaasAhsan commented on core/shared/src/main/scala/cats/effect/IO.scala:
```diff
@@ -277,6 +278,53 @@ object IO extends IOCompanionPlatform with IOLowPriorityImplicits {
+      error <- Deferred[IO, Throwable]
+      queue <- Queue.bounded[IO, (Deferred[IO, B], IO[B])](n)
+      pairs <- ta.traverse(a => Deferred[IO, B].map(p => (p, f(a))))
+      _ <- pairs.traverse_(queue.offer(_))
```
Should we `parTraverse` here to make things more nondeterministic?

SystemFw replied:
I think the whole point of this implementation is to not do that, otherwise you can use the implementation with `Semaphore`

SystemFw commented:
https://gitter.im/typelevel/cats-effect-dev?at=5f9838d1a0a3072b43aad96c

Some relevant discussion happening here

TimWSpence committed:
Merge remote-tracking branch 'upstream/series/3.x' into port-par-traverse-n

TimWSpence committed:
Merge remote-tracking branch 'upstream/series/3.x' into port-par-traverse-n

TimWSpence committed:
Re-implement par{Sequence, Traverse}N for GenConcurrent

Requires no dependency on Semaphore or Queue from std as we don't
want to move them into kernel

TimWSpence committed:
Scalafmt

TimWSpence committed:
Merge remote-tracking branch 'upstream/series/3.x' into port-par-traverse-n

TimWSpence committed:
Add syntax specs

TimWSpence committed:
Test that parTraverseN is cancelable

TimWSpence committed:
Handle errors in par{Sequence, Traverse}N

TimWSpence committed:
Initial draft of mini semaphore for parTraverseN

Prevents windowing, where the first n elements of the
traversable structure must be run first in
parTraverseN

TimWSpence committed:
Spec for mini semaphore

TimWSpence committed:
Fix parTraverseN spec

SystemFw commented:
@TimWSpence if we get this to build, I will merge and do the changes on MiniSemaphore in a separate PR :)

djspiewak commented:
Firefox needs a more generous timeout, in my experience. Usually about twice as high. Btw you can run the Firefox tests locally by installing `geckodriver` (homebrew is sufficient for this if you're on a mac) and just running `ciFirefox`. Or alternatively, `; project coreJS; set useFirefox := true; test`

TimWSpence commented:
> Firefox needs a more generous timeout, in my experience. Usually about twice as high. Btw you can run the Firefox tests locally by installing `geckodriver` (homebrew is sufficient for this if you're on a mac) and just running `ciFirefox`. Or alternatively, `; project coreJS; set useFirefox := true; test`

Where is the timeout configured for firefox sorry? Or why has this not failed before?

TimWSpence committed:
Greater timeout for property tests

TimWSpence commented:
> > Firefox needs a more generous timeout, in my experience. Usually about twice as high. Btw you can run the Firefox tests locally by installing `geckodriver` (homebrew is sufficient for this if you're on a mac) and just running `ciFirefox`. Or alternatively, `; project coreJS; set useFirefox := true; test`
> 
> Where is the timeout configured for firefox sorry? Or why has this not failed before?

Never mind - I realized it's probably because we now have property tests that run lots of iterations. I've moved them to a separate test suite with a larger timeout
